spock-specs/src/main/groovy/org/spockframework/specs/extension/SnapshotExtension.groovy

```diff
@@ -17,24 +21,63 @@ class SnapshotExtension implements IAnnotationDrivenExtension<Snapshot> {
   }
 
   @Override
-  void visitFieldAnnotation(Snapshot annotation, FieldInfo field) {
-    Assert.that(field.type.isAssignableFrom(Snapshotter), "Field must be of type Snapshotter")
-    field.parent.bottomSpec.addSetupInterceptor(new SnapshotInterceptor(annotation, field))
+  void visitSpecAnnotation(Snapshot annotation, SpecInfo specInfo) {
+    specInfo.features.each { visitFeatureAnnotation(annotation, it) }
+  }
+
+  @Override
+  void visitFeatureAnnotation(Snapshot annotation, FeatureInfo feature) {
+    def snapshotterFields = feature.spec.allFields.findAll {
+      ((it.type == Snapshotter) || (it.type.isAssignableFrom(Snapshotter) && (it.name == 'snapshotter'))) &&
+        !it.static && !it.shared
+    }
+    def snapshotterParameters = feature.featureMethod.parameters.findAll {
+      ((it.reflection.type == Snapshotter) || (it.reflection.type.isAssignableFrom(Snapshotter) && (it.name == 'snapshotter'))) &&
+        !(it.name in feature.dataVariables)
+    }
+    Assert.that(
+      (snapshotterFields.size() != 0) || (snapshotterParameters.size() != 0),
+      'There must be at least one Snapshotter parameter that is not a data variable or one Snapshotter field that is neither static nor @Shared'
+    )
+
+    feature.addIterationInterceptor(new SnapshotFieldInterceptor(annotation, snapshotterFields))
+    feature.featureMethod.addInterceptor(new SnapshotParameterInterceptor(annotation, snapshotterParameters))
+  }
+
+  @CompileStatic
+  class SnapshotFieldInterceptor implements IMethodInterceptor {
+    private final List<FieldInfo> fields
+    private final Snapshot annotation
+
+    SnapshotFieldInterceptor(Snapshot annotation, List<FieldInfo> fields) {
+      this.fields = fields
+      this.annotation = annotation
+    }
+
+    @Override
+    void intercept(IMethodInvocation invocation) throws Throwable {
+      fields
+        .findAll { it.readValue(invocation.instance) == null }
+        .each { it.writeValue(invocation.instance, new Snapshotter(invocation.iteration, config.rootPath, config.updateSnapshots, annotation.extension())) }
+      invocation.proceed()
+    }
   }
 
   @CompileStatic
-  class SnapshotInterceptor implements IMethodInterceptor {
-    private final FieldInfo field
+  class SnapshotParameterInterceptor implements IMethodInterceptor {
+    private final List<ParameterInfo> parameters
     private final Snapshot annotation
 
-    SnapshotInterceptor(Snapshot annotation, FieldInfo field) {
-      this.field = field
+    SnapshotParameterInterceptor(Snapshot annotation, List<ParameterInfo> parameters) {
+      this.parameters = parameters
       this.annotation = annotation
     }
 
     @Override
     void intercept(IMethodInvocation invocation) throws Throwable {
-      field.writeValue(invocation.instance, new Snapshotter(invocation.method.iteration, config.rootPath, config.updateSnapshots, annotation.extension()))
+      parameters
```

Comment:
:x: Take a look at #1599 use the new `invocation.resolve` method instead, or just replace this with the `ParameterResolver.Interceptor`.

Reply:
I had a look.
Besides that this new method is imho misbehaving (will open a separate issue to discuss it), that would change semantic, depending on the outcome on the conversation above regarding "Shouldn't this be an error", this semantic change might be good or bad.